loopy/symbolic.py

```diff
@@ -358,132 +376,154 @@ class UncachedWalkMapper(WalkMapperMixin[P], UncachedWalkMapperBase[P]):
                     self.rec(expr.index, PREC_NONE)),
                 enclosing_prec, PREC_CALL)
 
-    def map_loopy_function_identifier(self, expr, enclosing_prec):
-        return "{}<{}>".format(
-                type(expr).__name__,
-                ", ".join(str(a) for a in expr.__getinitargs__()))
+    def map_loopy_function_identifier(
+            self, expr: FunctionIdentifier, enclosing_prec: int) -> str:
+        from dataclasses import fields
+        initargs = ", ".join(f"{getattr(expr, fld.name)}" for fld in fields(expr))
+
+        return f"{type(expr).__name__}<{initargs}>"
 
-    def map_rule_argument(self, expr, enclosing_prec):
-        return "<arg%d>" % expr.index
+    def map_rule_argument(self, expr: RuleArgument, enclosing_prec: int) -> str:
+        return f"<arg{expr.index}>"
 
-    def map_type_cast(self, expr, enclosing_prec):
+    def map_type_cast(self, expr: TypeCast, enclosing_prec: int) -> str:
         from pymbolic.mapper.stringifier import PREC_NONE
         child = self.rec(expr.child, PREC_NONE)
         return f"cast({expr.type!r}, {child})"
 
-    def map_resolved_function(self, expr, prec):
+    def map_resolved_function(self, expr: ResolvedFunction, enclosing_prec: int) -> str:
         # underlining a resolved call
         return "\u0332".join(str(expr.function))
 
-    def map_sub_array_ref(self, expr, prec):
-        return "[{inames}]: {subscr}".format(
-                inames=",".join(self.rec(iname, prec) for iname in
-                    expr.swept_inames),
-                subscr=self.rec(expr.subscript, prec))
+    def map_sub_array_ref(self, expr: SubArrayRef, enclosing_prec: int) -> str:
+        inames = ",".join(self.rec(iname, enclosing_prec)
+                          for iname in expr.swept_inames)
+        subscr = self.rec(expr.subscript, enclosing_prec)
 
-    def map_fortran_division(self, expr, enclosing_prec):
+        return f"[{inames}]: {subscr}"
+
+    def map_fortran_division(self, expr: FortranDivision, enclosing_prec: int) -> str:
         from pymbolic.mapper.stringifier import PREC_NONE
-        result = self.map_quotient(expr, PREC_NONE)
+        result = self.map_quotient(expr, PREC_NONE)  # pyright: ignore[reportArgumentType]
         return f"[FORTRANDIV]({result})"
 
 
 class UnidirectionalUnifier(UnidirectionalUnifierBase):
-    def map_reduction(self, expr, other, unis):
+    def map_reduction(
+                self,
+                expr: Reduction,
+                other: Expression,
+                urecs: Sequence[UnificationRecord]) -> Sequence[UnificationRecord]:
         if not isinstance(other, type(expr)):
-            return self.treat_mismatch(expr, other, unis)
+            return self.treat_mismatch(expr, other, urecs)
+
         if (expr.inames != other.inames
-                or type(expr.function) is not type(other.function)):
+                or type(expr.operation) is not type(other.operation)):
```

Comment:
This also doesn't look particularly used :grin:

Reply:
...tested, you mean? I'd still like to keep it.